datafusion/expr/src/type_coercion/binary.rs

```diff
@@ -575,6 +575,7 @@ fn temporal_coercion(lhs_type: &DataType, rhs_type: &DataType) -> Option<DataTyp
         },
         (Timestamp(_, tz), Utf8) => Some(Timestamp(TimeUnit::Nanosecond, tz.clone())),
         (Utf8, Timestamp(_, tz)) => Some(Timestamp(TimeUnit::Nanosecond, tz.clone())),
+        // TODO: need to investigate the result type for the comparison between timestamp and date
```

Comment:
https://github.com/apache/arrow-datafusion/issues/4644#issuecomment-1365869161
we need to get the behavior for the comparison in the PG.

Reply:
@liukun4515  comparison is usually boolean type, did you mean what is the common datatype for casting between timestamp and date in diff scenarios?

Reply:
Yes
the result of comparison is bool type, but if left and right are the same data type, we should get the common type for left and right through the `type coercion`.
For example, INT32>INT64, the common type is INT64, and we should cast the left to INT64.

Reply:
The type of `type coercion` is decided by `input types` and `op`.

Reply:
> @liukun4515 comparison is usually boolean type, did you mean what is the common datatype for casting between timestamp and date in diff scenarios?

I think it's not about `casting` from one type to other type, it is related the `type inference` for different input types and different operation or functions.

Reply:
I try the comparison timestamp with date in the Spark, and find the coerced type is timestamp.

```
spark-sql> explain extended select now()>cast('2000-01-01' as date);
== Parsed Logical Plan ==
'Project [unresolvedalias(('now() > cast(2000-01-01 as date)), None)]
+- OneRowRelation

== Analyzed Logical Plan ==
(now() > CAST(2000-01-01 AS DATE)): boolean
Project [(now() > cast(cast(2000-01-01 as date) as timestamp)) AS (now() > CAST(2000-01-01 AS DATE))#49]
+- OneRowRelation

== Optimized Logical Plan ==
Project [true AS (now() > CAST(2000-01-01 AS DATE))#49]
+- OneRowRelation

== Physical Plan ==
*(1) Project [true AS (now() > CAST(2000-01-01 AS DATE))#49]
+- *(1) Scan OneRowRelation[]
```


cc @comphead @alamb 

But current implementation of coerced type is date, we can fix this in the follow up PR.

Reply:
https://github.com/apache/arrow-datafusion/issues/4761

Reply:
Timestamp is probably consistent with other coercions that ensure no data is truncated (so going Date --> Timestamp is good as the the Timestamp could  be losslessly converted back to a Date). However, going from Timestamp --> Date would truncate the precision of the original timesteamp,